### PR TITLE
[FIRRTL] Prerequisites for Layer-Associated Probes

### PR DESCRIPTION
**Pull request by seldridge**
Miscellaneous fixes to get layer-associated probes working.

This primarily makes it so that you can use a probe in a layerblock to define something outside it.  Checks are added along the lines of what is suggested in https://github.com/llvm/circt/pull/6551#issuecomment-1881401430 to do verification of layers. There are only two types of checks that are needed to verify symbols:

1. Module ports are verified that any layer-associated probe points at a defined layer.
2. Ref Cast ops are verified that they cast to a legal destination.
3. Wires are verified that they cast to a defined layer.

There are a lot of other checks that could be added, but are unnecessary. E.g., refsend and refsub don't need to be checked because the only way to get a probe referring to an undefined destination is through a module, refcast, or wire. Hence, these unnecessary checks are skipped.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 1]
```diff
@@ -1455,6 +1455,17 @@ static LogicalResult verifyPortSymbolUses(FModuleLike module,
   // verify types in ports.
   for (size_t i = 0, e = module.getNumPorts(); i < e; ++i) {
     auto type = module.getPortType(i);
+
+    if (auto refType = dyn_cast<RefType>(type)) {
+      auto layer = refType.getLayer();
+      if (layer && !dyn_cast_or_null<LayerOp>(
+                       symbolTable.lookupSymbolIn(circuitOp, layer)))
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1462]: `symbolTable.lookupSymbolIn<LayerOp>` does the `dyn_cast_or_null` for you.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 2]
```diff
@@ -3045,6 +3056,20 @@ void WireOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
...
 std::optional<size_t> WireOp::getTargetResultIndex() { return 0; }
 
+LogicalResult WireOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto circuitOp = getOperation()->getParentOfType<CircuitOp>();
+  auto refType = dyn_cast<RefType>(getType(0));
+  if (!refType)
+    return success();
+
+  auto layer = refType.getLayer();
+  if (layer && !symbolTable.lookupSymbolIn(circuitOp, layer))
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:3066]:
```suggestion
  if (layer && !symbolTable.lookupSymbolIn<LayerOp>(circuitOp, layer))
```
?
(to match the other)

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 3]
```diff
@@ -3045,6 +3056,20 @@ void WireOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
 
 std::optional<size_t> WireOp::getTargetResultIndex() { return 0; }
 
+LogicalResult WireOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto circuitOp = getOperation()->getParentOfType<CircuitOp>();
+  auto refType = dyn_cast<RefType>(getType(0));
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:3061]: It's probably fine but hardcoding operand numbers is less readable and a maintenance burden.

Should this be `firrtl::dyn_cast` in case we ever support aliases to non-base?

Maybe this should recursively check the type for any references? Although post-LOA this does the trick, but making sure that's intentional.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 4]
```diff
@@ -5691,13 +5716,64 @@ FIRRTLType RefSubOp::inferReturnType(ValueRange operands,
...
+                           layerPointer.getNestedReferences().drop_back());
+  }
+
+  return success();
+}
+
+LogicalResult RefCastOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto circuitOp = getOperation()->getParentOfType<CircuitOp>();
+  auto layer = getType().getLayer();
+  if (layer && !symbolTable.lookupSymbolIn(circuitOp, layer))
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:5770]:
```suggestion
  if (layer && !symbolTable.lookupSymbolIn<LayerOp>(circuitOp, layer))
```

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 3 above)
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:3061]: Oh, right, need to fix `Forceable` for non-base things in Wires for that API to make sense again (`getDataType()`), blargh.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 5]
```diff
@@ -5691,13 +5716,64 @@ FIRRTLType RefSubOp::inferReturnType(ValueRange operands,
...
 
   return emitInferRetTypeError(
       loc, "ref.sub op requires a RefType of vector or bundle base type");
 }
 
+LogicalResult RefCastOp::verify() {
+
+  SymbolRefAttr layer;
+  auto layerBlockOp =
+      dyn_cast<LayerBlockOp>((*this)->getBlock()->getParentOp());
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:5730]: Can layers have region-creating ops like whens or match statements?

**Review by dtzSiFive — COMMENTED** [review]
Some comments, will revisit, need to think some more.

>  because the only way to get a probe referring to an undefined destination is through a module, refcast, or wire. Hence, these unnecessary checks are skipped.

How can we capture this reasoning / assertion / property so it's true in the future as well?

What does checking a wire get you -- how does a wire get the probe if not from a ref.cast or module port?

(Looks like probe-generation is not layer-aware in this design direction, and so ref.cast "adds" it?)

**Comment by seldridge** [comment]
> What does checking a wire get you -- how does a wire get the probe if not from a ref.cast or module port?

It lets you reject this outright: https://github.com/llvm/circt/pull/6574/files#diff-e012a404f9fde3c2af947cbfbf8abeb5c8126f24a6f45d44d530e4923edb0510R1946-R1951

Yes, this would be checked if this was a legal FIRRTL circuit (it is illegal because the wire is not initialized). However, initialization checking is not something that can be checked immediately. Also, this circuit could be legally optimized to remove the wire because it has no uses. (Optimizations shouldn't make illegal circuits legal.)

Basically, this is trying to check only the ops which can create new layer-associated probes. I _think_ this is only possible for modules, ref.cast, and wire. WDYT?

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 5 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:5730]: They can. This looks like a bug in that case. It should walk backwards until it finds a layerblock (or bottoms out at a module). Good catch.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 1 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: Fixed in https://github.com/llvm/circt/pull/6574/commits/68b4e01b15af105b51776e0225616d1cf8853e38.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 2 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: Fixed in https://github.com/llvm/circt/pull/6574/commits/68b4e01b15af105b51776e0225616d1cf8853e38.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 4 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: Fixed in https://github.com/llvm/circt/pull/6574/commits/68b4e01b15af105b51776e0225616d1cf8853e38.

**Comment by seldridge** [comment]
https://github.com/llvm/circt/pull/6574/commits/dc9a1b8cdead59b3d19c2f2ba725e5043ebc2131 removes symbol users from `firrtl.ref.cast` op. This is unnecessary as the verifier to check the op seems to run first. This verifies that a legal destination is used. However, in order for that to succeed this needs to be in a layerblock with a legal `SymbolRefAttr`. I.e., there is no way to construct a `ref.cast` op that casts to a probe associated with a layer which does not exist or points at a non-layer symbol.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 5 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: Fixed in 54b7da58e.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 3 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: Right, `getDataType()` is approximately what I want here. However, I need the actual type, not the ref-stripped type.

I thought about the "recursive check" but opted not to as that only comes up for open aggregates and those are immediately factored out by LOA.

I switched over all the `dyn_cast` to use `type_dyn_cast` (which is what I think you were referring to?) in 63cd555df.

Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: Yeah, I mean `type_dyn_cast`, thanks.

We can make working with declarations and their storage/component types better separately, that's what `getDataType()` is supposed to be (it doesn't strip anything, it's as opposed to the type of the rwprobe result for Forceable's) but was borked and not yet fixed when Forceable's now can have things that aren't FIRRTLBaseType.

**Comment by seldridge** [comment]
The exact requirements for layer op verifiers are in this FIRRTL spec PR: https://github.com/chipsalliance/firrtl-spec/pull/160 This were defined after some offline iteration with @rwy7. Basically, a probe can be used as a "source" if it is visible to you. A probe may only be used as a "destination" (of a define, force, etc.) if it has the same color as the layer block (or is uncolored if not in a layer block).

This PR should already be implementing the "source" checks correctly. However, it is not implementing the "destination" checks correctly. Currently, the "destination" checks are the same as the "source" checks.

**Comment by rwy7** [comment]
I'm going to merge this PR as-is, and update the verifiers in a followup PR.
